descartes_ikfast/include/descartes_ikfast/impl/ikfast_kinematics.hpp

```diff
@@ -0,0 +1,216 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef DESCARTES_IKFAST_IKFAST_KINEMATICS_HPP
+#define DESCARTES_IKFAST_IKFAST_KINEMATICS_HPP
+
+#define IKFAST_HAS_LIBRARY
+#define IKFAST_NO_MAIN
+#include <descartes_ikfast/external/ikfast.h>
+#include <descartes_ikfast/ikfast_kinematics.h>
+#include <descartes_light/utils.h>
+#include <console_bridge/console.h>
+
+
+namespace descartes_light
+{
+
+template<typename FloatType>
+IKFastKinematics<FloatType>::IKFastKinematics(const Eigen::Transform<FloatType, 3, Eigen::Isometry> &world_to_robot_base,
+                                              const Eigen::Transform<FloatType, 3, Eigen::Isometry> &tool0_to_tip,
+                                              const IsValidFn<FloatType>& is_valid_fn,
+                                              const GetRedundantSolutionsFn<FloatType>& redundant_sol_fn)
+  : world_to_robot_base_(world_to_robot_base)
+  , tool0_to_tip_(tool0_to_tip)
+  , is_valid_fn_(is_valid_fn)
+  , redundant_sol_fn_(redundant_sol_fn)
+{
+}
+
+template<typename FloatType>
+bool IKFastKinematics<FloatType>::ik(const Eigen::Transform<FloatType, 3, Eigen::Isometry> &p,
+                                           std::vector<FloatType> &solution_set) const
+{
+  return IKFastKinematics<FloatType>::ik(p, is_valid_fn_, redundant_sol_fn_, solution_set);
+}
+
+template<typename FloatType>
+bool IKFastKinematics<FloatType>::ik(const Eigen::Transform<FloatType, 3, Eigen::Isometry>& p,
+                                     const IsValidFn<FloatType>& is_valid_fn,
+                                     const GetRedundantSolutionsFn<FloatType>& redundant_sol_fn,
+                                     std::vector<FloatType>& solution_set) const
+{
+  const Eigen::Transform<FloatType, 3, Eigen::Isometry> in_robot = world_to_robot_base_.inverse() * p * tool0_to_tip_.inverse();
+
+  // Convert to ikfast data type
+  Eigen::Transform<IkReal, 3, Eigen::Isometry> ikfast_tcp = in_robot.template cast<IkReal>();
+
+  // Decompose
+  const Eigen::Matrix<IkReal, 3, 1> translation = ikfast_tcp.translation();
+
+  // Note the row major ordering here: IkFast expects the matrix in r00, r01, r02, ..., r11, r12, r13
+  // ordering
+  const Eigen::Matrix<IkReal, 3, 3, Eigen::RowMajor> rotation = ikfast_tcp.rotation();
+
+  // Call IK (TODO: Make a better solution list class? One that uses vector instead of list)
+  ikfast::IkSolutionList<IkReal> ikfast_solution_set;
+  ComputeIk(translation.data(), rotation.data(), nullptr, ikfast_solution_set);
+
+  // Unpack the solutions into the output vector
+  const auto n_sols = ikfast_solution_set.GetNumSolutions();
+  int ikfast_dof = dof();
+
+  std::vector<IkReal> ikfast_output;
+  ikfast_output.resize(n_sols * ikfast_dof);
+
+  for (std::size_t i = 0; i < n_sols; ++i)
+  {
+    const auto& sol = ikfast_solution_set.GetSolution(i);  // I hate ikfast - this actually walks the list EVERY time from
```

Comment:
Consider removing the comment, despite your feelings toward IKFast

Reply:
I cannot take credit for this, but I will remove it.

Reply:
I reworded it.